### PR TITLE
TTP: Test payment collection flow

### LINKED ISSUE #8370: [Mobile Payments] Start payment flow from the TTP explanation screen (M2)
- [ ] Create an order before starting the test flow
- [ ] Navigate to order details with created order

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #8370

### Description

The PR implements the "try a TTP payment" flow. It's based on Simple payments flow, but navigates to the order details after successful payments, so a user can make a refund

### Testing instructions

Test that there is no regression:
* Simple payments
* IPP
* TTP

The main flow happy path:
* Open payments hub
* Click TTP
* Click "try a payment"
* Perform TTP 
* Issue a refund

Try different cases when:
* A user cancels the flow in the middle
* Error happens in one of the operations (payment, refund)

### Images/gif

https://user-images.githubusercontent.com/4923871/221840686-205ad2ec-5594-4aac-bafc-ffc47ff985cd.mp4

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by lmischner** [comment]
Outside of the design question in Slack, I only found one issue I think we ought to fix:
* If a user has previously connected to a bluetooth reader and does the trial payment workflow, the bluetooth reader is still connected (and so they can't easily try tap to pay without first going back and disconnecting the bluetooth reader). Can we disconnect from bluetooth readers on tapping "Try a payment"?

I view these as lower priority / if time: 
* In the Issue Refund workflow, can we automatically toggle the "refund fees" option to save the user from needing to tap?
* Will merchants be frustrated that these trial payments show up as actual transactions in their orders screen? It seems helpful to have the orders there in case they get interrupted while going through the trial, though, and still need to issue their refund. Maybe we could update the customer name from "Guest" to "Tap to Pay Trial Payment" or something to make it clearer?
* It also feels odd that I get a notification for the order from the trial payment workflow.

Coverage:
* TTP through trial payment, simple payments, and collect payment from order workflows
* Bluetooth reader from simple payments + order workflows
* TTP trial with Stripe + WC Payments
* In trial payment workflow, cancel flow by disconnecting from wifi and turning off NFC setting. There's probably more to do in this area around other types of errors, especially since I'm not as familiar with ones that could occur during the refund workflow.

**Comment by kidinov** [comment]
@lmischner Thank you for taking a look!

> If a user has previously connected to a bluetooth reader and does the trial payment workflow, the bluetooth reader is still connected (and so they can't easily try tap to pay without first going back and disconnecting the bluetooth reader). Can we disconnect from bluetooth readers on tapping "Try a payment"?

Good catch! That will be fixed in [this PR](https://github.com/woocommerce/woocommerce-android/pull/8435). I check if the selected reader type matches the one that is connected and if not then I do disconnection

> In the Issue Refund workflow, can we automatically toggle the "refund fees" option to save the user from needing to tap?

It sounds like a good improvement, but I noticed the same in iOS, and I wonder if this is something we are missing. @AnirudhBhat maybe you have insight on this as you've worked on interac refunds?

> Will merchants be frustrated that these trial payments show up as actual transactions in their orders screen? It seems helpful to have the orders there in case they get interrupted while going through the trial, though, and still need to issue their refund. Maybe we could update the customer name from "Guest" to "Tap to Pay Trial Payment" or something to make it clearer?
> It also feels odd that I get a notification for the order from the trial payment workflow.

Those are very good points. I checked and looks like first, we need to create a customer, and then when we create an order we already should use the ID of this customer, so it's quite an effort. We can a customer note to the order and it's relatively simple. @joshheald wdyt? Is this worth it?

I checked briefly our options, and I don't see how we can disable notification for a specific order 🙁

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodViewModel.kt [hunk 1]
```diff
@@ -324,14 +325,15 @@ class SelectPaymentMethodViewModel @Inject constructor(
...
+                TRY_TAP_TO_PAY -> NavigateToOrderDetails(cardReaderPaymentFlowParam.orderId)
                 ORDER -> NavigateBackToOrderList
             }
         )
     }
 
     private fun Payment.toAnalyticsFlowParams() =
         AnalyticsTracker.KEY_FLOW to when (paymentType) {
-            SIMPLE -> AnalyticsTracker.VALUE_SIMPLE_PAYMENTS_FLOW
+            SIMPLE, TRY_TAP_TO_PAY -> AnalyticsTracker.VALUE_SIMPLE_PAYMENTS_FLOW
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodViewModel.kt:336]: ❓ Just wanted to double-check if this behavior is intended. We are logging `TRY_TAP_TO_PAY` as "simple_payment" in the analytics.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodViewModel.kt [hunk 2]
```diff
@@ -159,7 +160,7 @@ class SelectPaymentMethodViewModel @Inject constructor(
             )
         )
         val messageIdForPaymentType = when (cardReaderPaymentFlowParam.paymentType) {
-            SIMPLE -> R.string.simple_payments_cash_dlg_message
+            SIMPLE, TRY_TAP_TO_PAY -> R.string.simple_payments_cash_dlg_message
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodViewModel.kt:163]: ❓ What do you think about writing a test to verify proper `messageIdForPaymentType` is being displayed when `TRY_TAP_TO_PAY` is the payment type?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodViewModel.kt (hunk 1 above)
Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodViewModel.kt:336]: Also, we could write a test to verify this behavior. WDYT?

**Review by AnirudhBhat — APPROVED** [review]
Thanks for your work on this @kidinov. It works as expected. I've left a couple of minor comments that are non-blocking for the merge. Please have a look.

One thing I noticed is that even after tapping "try tap to payment". We get prompted to choose the type of reader. I feel this is a bit odd. Since this is a tutorial, I think it would be better if we make it as easy and as simple as possible for the users. Removing that dialog makes it easier, avoids confusion, and also one less click.

I noticed that we are already passing the "TRY_TAP_TO_PAY" as a payment type argument to the `CardReaderTypeSelectionViewModel`. We could probably detect this and send users directly to TTP flow in this case. WDYT?

btw, this is just a minor nice-to-have thing.

**Comment by AnirudhBhat** [comment]
> @AnirudhBhat maybe you have insight on this as you've worked on interac refunds? 

AFAICS, I don't see any harm in turning it on automatically for this flow.

**Comment by lmischner** [comment]
> I noticed that we are already passing the "TRY_TAP_TO_PAY" as a payment type argument to the CardReaderTypeSelectionViewModel. We could probably detect this and send users directly to TTP flow in this case. WDYT?

@AnirudhBhat, I'm torn on this. I agree it would feel more slick to just hop into the workflow, but we also don't want them to be surprised in a real life interaction by a popup they haven't seen before. Is there a way to show the popup but to disable the bluetooth reader button? Some other middle ground?

**Comment by AnirudhBhat** [comment]
>I'm torn on this. I agree it would feel more slick to just hop into the workflow, but we also don't want them to be surprised in a real life interaction by a popup they haven't seen before. Is there a way to show the popup but to disable the bluetooth reader button? Some other middle ground?

I understand. Disabling the Bluetooth reader button sounds good to me. We could also add a small tooltip that says "click here" above the "Tap to pay" button. I think adding a tooltip would require some effort.

With that said, I don't have a strong opinion on this. I'm fine with whatever we decide (even leaving it as it is now)

**Comment by joshheald** [comment]
> I checked and looks like first, we need to create a customer, and then when we create an order we already should use the ID of this customer, so it's quite an effort. We can a customer note to the order and it's relatively simple. @joshheald wdyt? Is this worth it?

@kidinov  I don't think it's worth a lot of effort, but AFAIK a Customer is simply a WP user with a very limited role. I was going to experiment with using the User ID for the logged in user on the order to see whether that worked: I think it should, and it would be worth that little bit of effort.

I do think the transactions should stay in the transaction list though, particularly because of the refund. We could consider also adding a customer/order note saying that this was a Tap to Pay test payment.

**Comment by joshheald** [comment]
> I noticed that we are already passing the "TRY_TAP_TO_PAY" as a payment type argument to the CardReaderTypeSelectionViewModel. We could probably detect this and send users directly to TTP flow in this case. WDYT?

@lmischner @AnirudhBhat Just to add the iOS perspective on this: merchants must be connected to the built-in reader before they get to the TTP test payment, as it comes after the "set-up" step, which is just a connection to the reader. That means they won't be prompted for a reader type.

I know that connection isn't required on Android, so it's extra work to skip the prompt, but I don't think we should show the reader selection prompt... fortunately... we won't have to, because of later work... read on!

>  I'm torn on this. I agree it would feel more slick to just hop into the workflow, but we also don't want them to be surprised in a real life interaction by a popup they haven't seen before.

@lmischner I totally get and share your concern here: the payment flow should be true to what they see in the real app.

Another part of M2 is to remove the reader selection pop-up, and instead show `Bluetooth reader` and `Tap to Pay` on the Payment Method screen, which we should leave in the test payment flow, as it is here already. They need to know not to use `Cash`, `Payment link`, or `Bluetooth reader`, but those things seem pretty obvious when `Tap to Pay` is on offer there.

Hope that addresses all concerns?

**Comment by kidinov** [comment]
> @kidinov I don't think it's worth a lot of effort, but AFAIK a Customer is simply a WP user with a very limited role. I was going to experiment with using the User ID for the logged in user on the order to see whether that worked: I think it should, and it would be worth that little bit of effort.


@joshheald 
Nice! Let me know how it goes, in meanwhile I created a ticket to improve this if needed - https://github.com/woocommerce/woocommerce-android/issues/8497

**Pull request merged**